Added pre-defined dimensions support; Added AWS unique id support

sheremetat opened the pull request.

unsetbit commented on lib/client/signal_fx_client.js:
```diff
@@ -19,26 +19,49 @@ var version = require('../../package.json').version;
   this.asyncRunning = false;
+
+  if (this.enableAmazonUniqueId) {
+    this._retrieveAWSUniqueId(function (isSuccess, AWSUniqueId) {
```
How is this (or is this) protected from a race condition in which SignalFxClient is instantiated and then immediately .send is called to send datapoint while the unique id is still being retrieved?

sheremetat replied:
Fixed, please review again

unsetbit commented on lib/client/signal_fx_client.js:
```diff
@@ -19,26 +19,55 @@ var version = require('../../package.json').version;
+  this.globalDimensions = params.dimensions || {};
+  this.enableAmazonUniqueId = params.enableAmazonUniqueId || false;
 
+  this.rawDatas = [];
```
'data' is already a plural, so this var should just be named rawData

unsetbit commented on lib/client/signal_fx_client.js:
```diff
@@ -134,6 +207,24 @@ SignalFxClient.prototype.startAsyncSend = function () {
 
+SignalFxClient.prototype.startAsyncEventSend = function () {
+  if (this.loadingAWSUniqueId) {
+    return;
```
Wouldn't this cause a .send() immediately following instantiation of the SignalFxClient to fail silently? It looks like this is just doing a noop if .send is called which the AWS unique id is being loaded. It should wait for the unique id to be loaded and then complete the .send request.

sheremetat commented:
Fixed code review issue. I have used promises to guaranty sending all data after AWS unique Id was loaded (or wasn't). Also I have removed async library from app.  Please review it again. Thank you.

unsetbit commented on lib/client/signal_fx_client.js:
```diff
@@ -48,14 +84,25 @@ SignalFxClient.prototype.API_ENDPOINT_SUFFIX = 'v1/event';
+  this.rawData.push(data);
+  return this.loadAWSUniqueId
+    .then(function () {
+      _this.processingData();
```
Since this doesn't return another promise, the following .then is unnecessary, it can be added to this block. If _this.processingData is a promise, it should be returned here.

unsetbit commented on lib/client/signal_fx_client.js:
```diff
@@ -109,29 +199,33 @@ SignalFxClient.prototype.getHeaderContentType = function () {
+  var dataToSend = _this._batchData(datapointsList);
+  if (dataToSend) {
+    var url = _this.ingestEndpoint + '/' + _this.INGEST_ENDPOINT_SUFFIX;
+    _this.post(dataToSend, url, _this.getHeaderContentType(), function () {
```
empty function here

sheremetat commented:
@unsetbit done, you can review it again
